### PR TITLE
Single alpha per layer. Defaults changed.

### PR DESCRIPTION
**Pull request by brohrer**
Playing with alpha, I found very strong results with 1 shared alpha value across the layer. Initial observations suggest that one alpha per kernel doesn't add anything to performance, while incurring the addition of extra parameters.

Inspecting the alpha values chosen by LSUV_like_init(), they fall roughly on the range from 10-50 for a small model I'm running on Fashion MNIST. Choosing an initial alpha of 10 and disabling auto init generates the strongest results I've seen from my model yet.

@ducha-aiki, what do you think about moving to a single alpha per layer and removing the LSUV init? You've spent the most time thinking about it, so it's possible you've built an intuition around it I don't have yet.

@4rtemi5 based on what I've seen, the alpha boosts performance noticeably. Based on what implementation we settle on in PyTorch I plan to mimic it in the Jax code. It *might* make a difference in getting to 90% on CIFAR-10.  

This diff is a little messy-it also has some changes to the demo scripts to make them compatible with the new API.

### CODE REVIEW

**Comment by ducha-aiki** [comment]
I am for making shared alpha per layer and making 10 as default value. 
I am against removing LSUV-like init, because it is already switched off by default, so no harm is done. However, when we speak about some deeper arch's, having a possibility of auto-init gives you flexibility and some safe option.v

**Comment by 4rtemi5** [comment]
Cool! Thanks for doing the legwork on this! 
Currently I'm checking the alpha param with your suggested init in my jax code but I have a question:
Is the scaling by alpha not undone when normalizing in the next layer? Or is the alpha only supposed to be used in the last layer before the dense output?

**Comment by brohrer** [comment]
@ducha-aiki, sounds good. I think the LSUV-like method will need a bit of modification (maybe a mean() across kernels' alpha estimates?) to adapt it to a single shared alpha per layer. But I agree, having a robust auto-init takes one more hyperparameter out of the mix, and I like that. I'll see if I can find that adaptation.

**Comment by brohrer** [comment]
@4rtemi5, I had the same intuition, but after playing around with it I found that alphas throughout improved performance compared to an alpha in the final layer. My best guess is that it's a numerical issue, and that especially at first it's easy for the
forward propagation pass to come up with almost all zeros, leaving the backprop pass very little to work with. The "sharp" nature of SCS means that it's highly selective and with randomly initialized kernels, there may not be much of the features it's looking for initially. The use of the q parameter serves to attenuate the forward signal too. It's possible that an alpha > 1 helps compensate for these. It makes sense that it would become necessary more often with more SCS layers.

**Comment by 4rtemi5** [comment]
@brohrer Still a bit sceptical honestly. I don't see any improved performance in my code and both runs I have launched have crashed with exploding gradients: 
image

**Comment by brohrer** [comment]
@4rtemi5 it looks like there's more going on than we understand, for sure. Do you have a gist showing your implementation and the model you're using it with? I'm wondering if there are any differences.

It feels like it's time to take a more careful look at the distribution of feature activities and parameter, and how they change by layer and over time during training. I think this will be my next little project.

**Comment by brohrer** [comment]
@ducha-aiki I added a mean() to get a single alpha initialization value per layer. 

@4rtemi5 there's enough we don't understand about this yet that I'm comfortable not changing the Jax implementation until we do. For example, I was seeing the PyTorch version break with an MNIST test that Jax handled fine. It may be something deep in how the calculations are performed that makes the difference.

I'll go ahead and merge this and we can make other changes from there.

**Comment by brohrer** [comment]
@4rtemi5 , another thought that just occurred to me - maybe it's not necessary to have alpha be learnable. If you feel like it, trying a hard-coded constant alpha of say 10 might prevent the catastrophic accuracy losses you're seeing.

**Comment by 4rtemi5** [comment]
Just did that half an hour ago and the crashes are gone but I still don't see any change in accuracy compared to the baseline. Really puzzled...

**Comment by brohrer** [comment]
It may be that Jax does something different in the fine details that makes it more numerically stable. :man_shrugging:
